app/internal/page/certificate_provider_details_test.go

```diff
@@ -0,0 +1,320 @@
+	resp := w.Result()
+
+	assert.Nil(t, err)
+	assert.Equal(t, http.StatusOK, resp.StatusCode)
+	mock.AssertExpectationsForObjects(t, template, dataStore)
+}
+
+func TestGetCertificateProviderDetailsWhenTemplateErrors(t *testing.T) {
+	w := httptest.NewRecorder()
+
+	dataStore := &mockDataStore{}
+	dataStore.
+		On("Get", mock.Anything, "session-id").
+		Return(nil)
+
+	template := &mockTemplate{}
+	template.
+		On("Func", w, &certificateProviderDetailsData{
+			App:  appData,
+			Form: &certificateProviderDetailsForm{},
+		}).
+		Return(expectedError)
+
+	r, _ := http.NewRequest(http.MethodGet, "/", nil)
+
+	err := CertificateProviderDetails(template.Func, dataStore)(appData, w, r)
+	resp := w.Result()
+
+	assert.Equal(t, expectedError, err)
+	assert.Equal(t, http.StatusOK, resp.StatusCode)
+	mock.AssertExpectationsForObjects(t, template, dataStore)
+}
+
+func TestPostCertificateProviderDetails(t *testing.T) {
+	w := httptest.NewRecorder()
+
+	dataStore := &mockDataStore{
+		data: Lpa{
+			CertificateProvider: CertificateProvider{
+				FirstNames: "John",
+			},
+		},
+	}
+	dataStore.
+		On("Get", mock.Anything, "session-id").
+		Return(nil)
+	dataStore.
+		On("Put", mock.Anything, "session-id", Lpa{
+			CertificateProvider: CertificateProvider{
+				FirstNames:  "John",
+				LastName:    "Doe",
+				Email:       "john@example.com",
+				DateOfBirth: time.Date(1990, time.January, 2, 0, 0, 0, 0, time.UTC),
+			},
+		}).
+		Return(nil)
+
+	form := url.Values{
+		"first-names":         {"John"},
+		"last-name":           {"Doe"},
+		"email":               {"john@example.com"},
+		"date-of-birth-day":   {"2"},
+		"date-of-birth-month": {"1"},
+		"date-of-birth-year":  {"1990"},
+	}
+
+	r, _ := http.NewRequest(http.MethodPost, "/", strings.NewReader(form.Encode()))
+	r.Header.Add("Content-Type", formUrlEncoded)
+
+	err := CertificateProviderDetails(nil, dataStore)(appData, w, r)
+	resp := w.Result()
+
+	assert.Nil(t, err)
+	assert.Equal(t, http.StatusFound, resp.StatusCode)
+	assert.Equal(t, taskListPath, resp.Header.Get("Location"))
+	mock.AssertExpectationsForObjects(t, dataStore)
+}
+
+func TestPostCertificateProviderDetailsWhenStoreErrors(t *testing.T) {
+	w := httptest.NewRecorder()
+
+	dataStore := &mockDataStore{}
+	dataStore.
+		On("Get", mock.Anything, "session-id").
+		Return(nil)
+	dataStore.
+		On("Put", mock.Anything, "session-id", mock.Anything).
+		Return(expectedError)
+
+	form := url.Values{
+		"first-names":         {"John"},
+		"last-name":           {"Doe"},
+		"email":               {"john@example.com"},
+		"date-of-birth-day":   {"2"},
+		"date-of-birth-month": {"1"},
+		"date-of-birth-year":  {"1990"},
+	}
+
+	r, _ := http.NewRequest(http.MethodPost, "/", strings.NewReader(form.Encode()))
+	r.Header.Add("Content-Type", formUrlEncoded)
+
+	err := CertificateProviderDetails(nil, dataStore)(appData, w, r)
+
+	assert.Equal(t, expectedError, err)
+	mock.AssertExpectationsForObjects(t, dataStore)
```

Comment:
Should we be checking for some kind of error status code here? Or are we holding this back for when we work on non-happy path?

Reply:
Because it's returning an error the thing above will set it to 500, but in this test it would still be 200 which would be confusing to check (I think). But like you say, this will all change when we actually figure out what we're going to do for non-happy path stuff